metafacture-io/src/main/java/org/metafacture/io/HttpOpener.java

```diff
@@ -117,21 +203,109 @@ public void setHeader(final String key, final String value) {
         headers.put(key.toLowerCase(), value);
     }
 
+    /**
+     * Sets the HTTP request method.
+     *
+     * @param method the request method
+     */
+    public void setMethod(final Method method) {
+        this.method = method;
+    }
+
+    /**
+     * Sets the HTTP request URL.
+     *
+     * @param url the request URL
+     */
+    public void setUrl(final String url) {
+        this.url = url;
+    }
+
     @Override
-    public void process(final String urlStr) {
+    public void process(final String input) {
         try {
-            final URL url = new URL(urlStr);
-            final URLConnection con = url.openConnection();
-            headers.forEach(con::addRequestProperty);
-            String enc = con.getContentEncoding();
-            if (enc == null) {
-                enc = headers.get(ENCODING_HEADER);
+            final String requestUrl = getInput(input, url);
+            final String requestBody = getInput(input,
+                    body == null && method.getRequestHasBody() ? INPUT_DESIGNATOR : body);
+
+            final HttpURLConnection connection =
+                (HttpURLConnection) new URL(requestUrl).openConnection();
+
+            connection.setRequestMethod(method.name());
+            headers.forEach(connection::addRequestProperty);
+
+            if (requestBody != null) {
+                connection.setDoOutput(true);
+                connection.getOutputStream().write(requestBody.getBytes());
             }
-            getReceiver().process(new InputStreamReader(con.getInputStream(), enc));
+
+            final InputStream errorStream = connection.getErrorStream();
+            final InputStream inputStream = errorStream != null ?
+                getErrorStream(errorStream) : getInputStream(connection);
+
+            final String contentEncoding = getEncoding(connection.getContentEncoding());
+            getReceiver().process(new InputStreamReader(inputStream, contentEncoding));
         }
         catch (final IOException e) {
             throw new MetafactureException(e);
         }
     }
 
+    private String getInput(final String input, final String value) {
+        final String result;
+
+        if (!INPUT_DESIGNATOR.equals(value)) {
+            result = value;
+        }
+        else if (inputUsed) {
+            result = null;
+        }
+        else {
+            inputUsed = true;
+            result = input;
+        }
+
+        return result;
+    }
+
+    private InputStream getInputStream(final HttpURLConnection connection) throws IOException {
+        try {
+            return connection.getInputStream();
+        }
+        catch (final IOException e) {
+            final int responseCode = connection.getResponseCode();
+            if (responseCode >= SUCCESS_CODE_MIN && responseCode <= SUCCESS_CODE_MAX) {
```

Comment:
Had to introduce error handling based on response code since `HttpUrlConnection` doesn't seem to behave as [expected](https://docs.oracle.com/javase/8/docs/api/java/net/HttpURLConnection.html#getErrorStream--):

> Returns the error stream if the connection failed but the server sent useful data nonetheless. The typical example is when an HTTP server responds with a 404, which will cause a FileNotFoundException to be thrown in connect, but the server sent an HTML help page with suggestions as to what to do.

Should we make the success code range configurable, as you did in #462?

Reply:
yeah, why not ? otah we could also do this later, if need arises.

Reply:
It's implemented but not tested yet.

Reply:
SUCCESS_CODE_MIN is configurable already?

Reply:
Both are, but I haven't pushed yet because I still need to write the tests.

Reply:
I think it's good to drop this now +1